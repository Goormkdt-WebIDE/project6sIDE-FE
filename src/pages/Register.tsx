import React, { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { useForm } from "react-hook-form";
import NameInput from "../components/form/NameInput";
import EmailInput from "../components/form/EmailInput";
import PasswordInput from "../components/form/PasswordInput";
import SubmitButton from "../components/form/SubmitButton";
import { AxiosError } from "axios";
import { FormValue } from "../service/http-requests/user-api";
import { notifyError, notifySuccess } from "../service/toast";
import { useAuthContext } from "../context/AuthContext";

function Register() {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormValue>();

  const { register: registerAPI } = useAuthContext();

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [errorFromSubmit, setErrorFromSubmit] = useState<string | null>(null);
  const [loading, setLoading] = useState(false);
  const navigate = useNavigate();

  const onSubmit = async (data: FormValue) => {
    try {
      setLoading(true);
      const response = await registerAPI(data);
      notifySuccess(response.data as string);
      navigate("/login");
    } catch (error) {
      const axiosError = error as AxiosError;
      const message = `회원가입에 실패했습니다. : ${axiosError.code}`;
      notifyError(message);
      setErrorFromSubmit(message);
      setLoading(false);
    }
  };

  return (
    <>
      <div className="relative min-h-screen bg-cover bg-center bg-opacity-25 flex items-center justify-center">
        <div className="text-center items-center w-full">
          <div className="flex flex-col items-center justify-center h-screen">
            <h1 className="text-blue-500 text-5xl pb-4 mb-2 border-none font-thin">
              6S IDE
            </h1>
            <h3 className="text-black text-3xl pb-4 mb-2 border-none font-thin">
              Register
            </h3>
            <form
              className="max-w-md w-full p-8 rounded-lg shadow-lg bg-opacity-90"
              onSubmit={handleSubmit(onSubmit)}
            >
              <NameInput<FormValue>
                register={register}
                errors={errors}
                name="name"
              />
              <EmailInput<FormValue>
                register={register}
                errors={errors}
                name="email"
              />
              <PasswordInput<FormValue>
                register={register}
                errors={errors}
                name="password"
              />
              <SubmitButton text="Register" loading={loading} />
              <Link to="/login" className="text-gray-600 text-sm mt-4">
                Already have an account? Sign in
              </Link>
            </form>
          </div>
        </div>
      </div>
    </>
  );
}

export default Register;
