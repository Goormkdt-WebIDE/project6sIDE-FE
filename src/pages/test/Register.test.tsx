import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import Register from "../Register";
import { ToastContainer } from "react-toastify";
import { withAuthContext, withRouter } from "../../test/utils";
import { Route } from "react-router-dom";

const loginPageParagraph = "로그인 페이지 입니다.";

describe("Register", () => {
  const fakeAuthObj = {
    register: jest.fn(),
    login: jest.fn(),
    passwordReset: jest.fn(),
  };

  afterEach(() => {
    fakeAuthObj.register.mockReset();
  });

  it("스냅샷 테스트", () => {
    const { asFragment } = renderRegister();
    expect(asFragment()).toMatchSnapshot();
  });

  it("회원가입 성공 시, login 페이지로 이동해야한다.", async () => {
    fakeAuthObj.register.mockImplementation(() =>
      Promise.resolve({ data: "회원가입성공" })
    );

    renderRegister();

    await userEvent.type(screen.getByPlaceholderText("Name"), "test");

    await userEvent.type(
      screen.getByPlaceholderText("Email"),
      "test@gmail.com"
    );

    await userEvent.type(screen.getByPlaceholderText("Password"), "password");

    await userEvent.click(screen.getByRole("button"));

    expect(await screen.findByText(loginPageParagraph)).toBeInTheDocument();
  });

  it("회원가입실패 시, 회원가입에 실패했다는 문구가 떠야한다", async () => {
    fakeAuthObj.register.mockImplementation(() => {
      throw new Error();
    });

    renderRegister();

    await userEvent.type(screen.getByPlaceholderText("Name"), "test");

    await userEvent.type(
      screen.getByPlaceholderText("Email"),
      "test@gmail.com"
    );

    await userEvent.type(screen.getByPlaceholderText("Password"), "password");

    await userEvent.click(screen.getByRole("button"));

    expect(
      await screen.findByText(/회원가입에 실패했습니다./i)
    ).toBeInTheDocument();
  });

  function Login() {
    return <p>{loginPageParagraph}</p>;
  }

  function renderRegister() {
    return render(
      withAuthContext(
        withRouter(
          <>
            <Route
              path="/register"
              element={
                <>
                  <Register />
                  <ToastContainer />
                </>
              }
            />
            <Route path="/login" element={<Login />}></Route>
          </>,
          "/register"
        ),
        fakeAuthObj
      )
    );
  }
});
