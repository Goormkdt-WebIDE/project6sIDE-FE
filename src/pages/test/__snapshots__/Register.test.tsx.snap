// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register 스냅샷 테스트 1`] = `
<DocumentFragment>
  <div
    class="relative min-h-screen bg-cover bg-center bg-opacity-25 flex items-center justify-center"
  >
    <div
      class="text-center items-center w-full"
    >
      <div
        class="flex flex-col items-center justify-center h-screen"
      >
        <h1
          class="text-blue-500 text-5xl pb-4 mb-2 border-none font-thin"
        >
          6S IDE
        </h1>
        <h3
          class="text-black text-3xl pb-4 mb-2 border-none font-thin"
        >
          Register
        </h3>
        <form
          class="max-w-md w-full p-8 rounded-lg shadow-lg bg-opacity-90"
        >
          <input
            class="border rounded-md p-2 w-full mt-4"
            name="name"
            placeholder="Name"
          />
          <input
            class="border-none rounded-md p-2 w-full mt-4"
            name="email"
            placeholder="Email"
            type="email"
          />
          <input
            class="border-none rounded-md p-2 w-full mt-4"
            name="password"
            placeholder="Password"
            type="password"
          />
          <button
            class="bg-blue-400 text-white text-uppercase border-none rounded-md p-2 w-full my-4 text-xl font-thin letter-spacing-2 hover:bg-blue-500 active:transform active:translate-y-3 active:border-transparent active:opacity-80 cursor-pointer"
            type="submit"
          >
            Register
          </button>
          <a
            class="text-gray-600 text-sm mt-4"
            href="/login"
          >
            Already have an account? Sign in
          </a>
        </form>
      </div>
    </div>
  </div>
  <div
    class="Toastify"
  />
</DocumentFragment>
`;
